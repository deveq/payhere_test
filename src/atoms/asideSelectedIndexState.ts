import {atom} from 'recoil';

export const asideSelectedIndex = atom({
    key: 'asideSelectedIndex',
    default: 0,
});