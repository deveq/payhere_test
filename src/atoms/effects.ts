import { AtomEffect } from 'recoil';

export const themeEffect: AtomEffect<string> = ({ onSet, setSelf }) => {
	// localStorage로부터 theme을 가져온다
	let savedValue = localStorage.getItem('theme');
	if (savedValue === null) {
		// 없을 경우 browser에 설정된 테마를 꺼내온다
		const isBrowserDardMode =
			window.matchMedia &&
			window.matchMedia('(prefers-color-scheme: dark)').matches;
		savedValue = isBrowserDardMode ? 'DARK' : 'LIGHT';
	}

	setSelf(savedValue); // 'DARK' | 'LIGHT'

	onSet((newValue, _, isReset) => {
		isReset
			? localStorage.removeItem('theme')
			: localStorage.setItem('theme', newValue);
	});
};

// export const recordEffect: AtomEffect<boolean> = ({ onSet, setSelf }) => {
// 	// localStorage로부터 record를 가져온다
// 	const savedValue = localStorage.getItem('record');
// 	if (savedValue !== null) {
// 		setSelf(JSON.parse(savedValue));
// 	}

// 	onSet((newValue, _, isReset) => {
// 		isReset
// 			? localStorage.removeItem('record')
// 			: localStorage.setItem('record', JSON.stringify(newValue));
// 	});
// };

// export const repositoriesEffect: AtomEffect<string[]> = ({
// 	onSet,
// 	setSelf,
// }) => {
// 	const savedValue = localStorage.getItem('repositories');
// 	if (savedValue !== null) {
// 		setSelf(JSON.parse(savedValue));
// 	}

// 	onSet((newValue, _, isReset) => {
// 		isReset
// 			? localStorage.removeItem('repositories')
// 			: localStorage.setItem('repositories', JSON.stringify(newValue));
// 	});
// };

export const localStorageEffect: <T>(key: string) => AtomEffect<T> =
	(key: string) =>
	({ setSelf, onSet }) => {
		const savedValue = localStorage.getItem(key);
		if (savedValue != null) {
			setSelf(JSON.parse(savedValue));
		}
		onSet((newValue, _, isReset) => {
			isReset
				? localStorage.removeItem(key)
				: localStorage.setItem(key, JSON.stringify(newValue));
		});
	};
