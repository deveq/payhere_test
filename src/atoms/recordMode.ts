import { atom, selector } from "recoil";

const storedValue = JSON.parse(localStorage.getItem('record') || 'true');

export const recordModeState = atom<boolean>({
    key: 'recordMode',
    default: storedValue
});