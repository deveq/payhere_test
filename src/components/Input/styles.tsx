import { Children } from 'react';
import styled from 'styled-components';
import { FiSearch } from 'react-icons/fi';
import { flexCenter } from 'styles/helper';

interface HistoryProps {
	show?: boolean;
}

export const Container = styled.form`
	${flexCenter};
	width: 250px;
	border-radius: 10px;
	margin-left: 20px;
	padding: 10px;
	border: 1px solid ${({ theme }) => theme.border3};
	background: ${({ theme }) => theme.bg_element1};
	position: relative;
`;

export const Input = styled.input`
	border: none;
	outline: none;
	flex: 1;
	font-size: 16px;
	color: ${({ theme }) => theme.text1};
	background: ${({ theme }) => theme.bg_element1};
`;

export const Icon = styled(FiSearch).attrs({
	'data-testid': 'icon'
})`
	width: 20px;
	height: 20px;
	color: ${({ theme }) => theme.text1};
	cursor: pointer;
	&:hover,
	&:active {
		opacity: 0.6;
	}
`;

export const History = styled.ul<HistoryProps>`
	display: ${({ show }) => (show ? 'flex' : 'none')};
	position: absolute;
	top: 45px;
	background: ${({ theme }) => theme.bg_element1};
	width: 100%;
	height: ${({ children }) => `${Children.count(children) * 40}`}px;
	border-radius: 10px;
	z-index: 10;
	box-shadow: ${({ theme }) => theme.boxShadow};
	padding: 10px 20px;
	box-sizing: border-box;
	flex-direction: column;
	justify-content: space-around;
`;

export const HistoryItem = styled.li`
	color: ${({ theme }) => theme.text1};
	font-size: 18px;
	display: flex;
	justify-content: space-between;
	max-width: 90%;
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow: hidden;
	&:hover {
		opacity: 0.6;
	}
	cursor: pointer;
`;
