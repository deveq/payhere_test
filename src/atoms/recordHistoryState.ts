import { atom, selector } from "recoil";
import {localStorageEffect} from './effects';

export const recordHistoryState = atom<string[]>({
    key: 'recordHistory',
    default: [],
    effects: [localStorageEffect('recordHistory')]
});