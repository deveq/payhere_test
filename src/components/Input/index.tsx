import {useState} from 'react';
import { ChangeEventHandler, FormEventHandler } from 'react';
import * as Styled from './styles';
import {recordHistoryState} from 'atoms';
import {useRecoilValue} from 'recoil';

interface InputProps {
	onChange?: ChangeEventHandler<HTMLInputElement>;
	onSubmit?: FormEventHandler<HTMLFormElement>;
	iconVisible?: boolean;
	value?: string;
	defaultValue?: string;
    placeholder?: string;
	onClickHistoryItem?: (text: string) => void;
}

const Input = ({
	iconVisible = true,
	onChange,
	defaultValue,
	value,
    placeholder,
	onSubmit,
	onClickHistoryItem,
}: InputProps) => {
	const [isFocused, setIsFocused] = useState(false);
	const recordHistory = useRecoilValue(recordHistoryState);
	return (
		<Styled.Container onSubmit={(e) => {
			e.preventDefault();
			if (onSubmit) onSubmit(e);
		}}>
			<Styled.Input
				type="text"
				onChange={onChange}
				value={value}
				defaultValue={defaultValue}
                placeholder={placeholder}
				autoComplete={'off'}
				onFocus={() => {
					setIsFocused(true);
				}}
				onBlur={() => {
					setTimeout(() => {
						setIsFocused(false);
					}, 100);
				}}
			/>
			{iconVisible && <Styled.Icon />}
			<Styled.History show={isFocused && recordHistory.length > 0}>
				{
					[...recordHistory].reverse().map(item => (
						<Styled.HistoryItem 
						onClick={() => {
							if (!onClickHistoryItem) return;
							onClickHistoryItem(item)
						}}
						key={item}>{item}</Styled.HistoryItem>
					))
				}
			</Styled.History>
		</Styled.Container>
	);
};

export default Input;
