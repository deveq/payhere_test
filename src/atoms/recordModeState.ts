import { atom, selector, DefaultValue } from "recoil";
import {localStorageEffect} from './effects';
import {recordHistoryState} from './recordHistoryState';

const recordModeState = atom<boolean>({
    key: 'recordMode',
    default: true,
    effects: [localStorageEffect('record')]
});

export const recordModeSelector = selector({
    key: 'recordModeSelector',
    get: ({get}) => get(recordModeState),
    set: ({set, reset}, newValue) => {
        if (newValue === false) {
            reset(recordHistoryState)
        }
        set(recordModeState,newValue);
    }
})

export {recordHistoryState};