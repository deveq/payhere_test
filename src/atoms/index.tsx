import {asideSelectedIndex} from './asideSelectedIndexState';
import {recordModeSelector, recordHistoryState} from './recordModeState';
import {repositoriesState} from './repositoriesState';
import {themeModeState} from './themeModeState';

export {
    asideSelectedIndex,
    recordModeSelector,
    repositoriesState,
    themeModeState,
    recordHistoryState
}