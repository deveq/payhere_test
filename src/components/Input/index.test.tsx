import {Children} from 'react';
import {screen, fireEvent} from '@testing-library/react';
import {renderWithWrapper} from 'tests';
import Input from './index';
import {recordHistoryState} from 'atoms';
import {useRecoilState} from 'recoil';

describe("<Input />", () => {
    it('정상적으로 렌더링 되는지 확인한다', () => {
        const {container} = renderWithWrapper(
            <Input />
        );

        // const form = screen.getByRole('form');
        const input = screen.getByRole('textbox');
        const form = input.parentElement;
        // const history = screen.findByRole('list');

        expect(input).toBeInTheDocument();
        expect(form).toBeInTheDocument();
        // expect(history).toBeInTheDocument();
        expect(container).toMatchSnapshot();

    })

    it('props에 따라 정상렌더링되는지 확인한다', () => {

        const {container} = renderWithWrapper(
            <Input
                iconVisible={true}
             />
        );
        
        const input = screen.getByRole('textbox');
        fireEvent.focus(input)
        const icon = screen.getByTestId('icon');
        expect(icon).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    })
})

// export const Container = styled.form`
// 	border: 1px solid ${({ theme }) => theme.border3};
// 	background: ${({ theme }) => theme.bg_element1};
// `;

// export const Input = styled.input`
// 	color: ${({ theme }) => theme.text1};
// 	background: ${({ theme }) => theme.bg_element1};
// `;

// export const Icon = styled(FiSearch)`
// 	color: ${({ theme }) => theme.text1};
// 	&:hover,
// 	&:active {
// 		opacity: 0.6;
// 	}
// `;

// export const History = styled.ul<HistoryProps>`
// 	display: ${({ show }) => (show ? 'flex' : 'none')};
// 	background: ${({ theme }) => theme.bg_element1};
// 	height: ${({ children }) => `${Children.count(children) * 40}`}px;
// 	box-shadow: ${({ theme }) => theme.boxShadow};
// `;

// export const HistoryItem = styled.li`
// 	color: ${({ theme }) => theme.text1};
// 	&:hover {
// 		opacity: 0.6;
// 	}
// `;
