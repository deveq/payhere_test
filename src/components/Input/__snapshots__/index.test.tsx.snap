// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> props에 따라 정상렌더링되는지 확인한다 1`] = `
<div>
  <form
    class="sc-bcXHqe hSmhPQ"
  >
    <input
      autocomplete="off"
      class="sc-gswNZR jTRvUa"
      type="text"
      value=""
    />
    <svg
      class="sc-dkrFOg cikoXo"
      data-testid="icon"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="11"
        cy="11"
        r="8"
      />
      <line
        x1="21"
        x2="16.65"
        y1="21"
        y2="16.65"
      />
    </svg>
    <ul
      class="sc-hLBbgP fGWmYr"
    />
  </form>
</div>
`;

exports[`<Input /> 정상적으로 렌더링 되는지 확인한다 1`] = `
<div>
  <form
    class="sc-bcXHqe hSmhPQ"
  >
    <input
      autocomplete="off"
      class="sc-gswNZR jTRvUa"
      type="text"
      value=""
    />
    <svg
      class="sc-dkrFOg cikoXo"
      data-testid="icon"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="11"
        cy="11"
        r="8"
      />
      <line
        x1="21"
        x2="16.65"
        y1="21"
        y2="16.65"
      />
    </svg>
    <ul
      class="sc-hLBbgP fGWmYr"
    />
  </form>
</div>
`;
